lib/vm/src/instance.rs

```diff
@@ -978,6 +1052,97 @@ impl InstanceHandle {
         Ok(handle)
     }
 
+    /// Create a new `InstanceHandle` pointing at the instance
+    /// pointed to by the given `VMContext` pointer.
+    ///
+    /// # Safety
+    /// This is unsafe because it doesn't work on just any `VMContext`, it must
+    /// be a `VMContext` allocated as part of an `Instance`.
+    pub(crate) unsafe fn from_vmctx(vmctx: *const VMContext) -> Self {
+        /*
+        let instance = (&*vmctx).instance();
+
+        Self {
+            instance: Arc::from_raw(instance as *const InstanceAllocator),
+            // We consider `vmctx` owns the `Instance`, not `Self`.
+            //owned_instance: false,
+        }
+        */
+        todo!()
+    }
+
+    /// Return a reference to the contained `Instance`.
+    pub(crate) fn instance(&self) -> &InstanceAllocator {
+        &self.instance
+    }
+
+    /// Get the locations of where the local `VMMemoryDefinition`s should be stored.
+    ///
+    /// This function lets us create `Memory` objects on the host with backing
+    /// memory in the VM.
+    ///
+    /// # Safety
+    /// - `instance_ptr` must point to enough memory that all of the offsets in
+    ///   `offsets` point to valid locations in memory.
+    pub unsafe fn memory_definition_locations(
+        instance_ptr: NonNull<u8>,
+        offsets: &VMOffsets,
+    ) -> Vec<NonNull<VMMemoryDefinition>> {
+        // `NonNull<u8>` here actually means `NonNull<Instance>`. See
+        // `Self::allocate_instance` to understand why.
+        let instance_ptr: NonNull<Instance> = instance_ptr.cast();
+
+        let num_memories = offsets.num_local_memories;
+        let num_memories = usize::try_from(num_memories).unwrap();
+        let mut out = Vec::with_capacity(num_memories);
+
+        let base_ptr = instance_ptr.as_ptr().add(mem::size_of::<Instance>());
+
+        for i in 0..num_memories {
+            let mem_offset = offsets.vmctx_vmmemory_definition(LocalMemoryIndex::new(i));
+            let mem_offset = usize::try_from(mem_offset).unwrap();
+
+            let new_ptr = NonNull::new_unchecked(base_ptr.add(mem_offset));
+
+            out.push(new_ptr.cast());
+        }
+
+        out
+    }
+
+    /// Get the locations of where the `VMTableDefinition`s should be stored.
+    ///
+    /// This function lets us create `Table` objects on the host with backing
+    /// memory in the VM.
+    ///
+    /// # Safety
+    /// - `instance_ptr` must point to enough memory that all of the offsets in
+    ///   `offsets` point to valid locations in memory.
+    pub unsafe fn table_definition_locations(
+        instance_ptr: NonNull<u8>,
+        offsets: &VMOffsets,
+    ) -> Vec<NonNull<VMTableDefinition>> {
+        // `NonNull<u8>` here actually means `NonNull<Instance>`. See
+        // `Self::allocate_instance` to understand why.
+        let instance_ptr: NonNull<Instance> = instance_ptr.cast();
```

Comment:
This also causes the pointer math below to be incorrect

Reply:
Yup, fixed!